inc/knowbaseitem_item.class.php

```diff
@@ -0,0 +1,396 @@
+
+      $massiveactionparams['ontop'] = false;
+      Html::showMassiveActions($massiveactionparams);
+
+      echo "</div>";
+      Html::printAjaxPager($type_name, $start, $number);
+   }
+
+   /**
+    * Displays linked dropdowns to add linked items
+    *
+    * @param CommonDBTM $item Item instance
+    * @param string     $name Field name
+    *
+    * @return string
+    */
+   static function dropdownAllTypes(CommonDBTM $item, $name) {
+      global $CFG_GLPI;
+
+      $onlyglobal = 0;
+      $entity_restrict = -1;
+      $checkright = true;
+
+      $rand = Dropdown::showAllItems(
+         $name,                 //select/@name
+         0,
+         0,
+         $entity_restrict,      //entity restriction
+         $CFG_GLPI['kb_types'], //types list
+         $onlyglobal,
+         $checkright            //checkright
+      );
+
+      return $rand;
+   }
+
+   /**
+    * Retrieve items for a knowbase item
+    *
+    * @param CommonDBTM $item      CommonDBTM object
+    * @param integer    $start     first line to retrieve (default 0)
+    * @param integer    $limit     max number of line to retrive (0 for all) (default 0)
+    * @param string     $sqlfilter to add an SQL filter (default '')
+    * @param boolean    $used      whether to retrieve data for "used" records
+    *
+    * @return array of linked items
+   **/
+   static function getItems(CommonDBTM $item, $start=0, $limit=0, $sqlfilter='', $used = false) {
+      global $DB;
+
+      $itemtype  = $item->getType();
+      $items_id  = $item->getField('id');
+      $itemtable = $item->getTable();
+
+      $select = $used === false ? '*' :
+          $item::getType() === KnowbaseItem::getType() ?
+             '`kb_linked`.`knowbaseitems_id`' :
+             '`kb_linked`.`items_id`';
+
+      $query = "SELECT `kb_linked`.*
+                FROM `glpi_knowbaseitems_items` as `kb_linked`
+                INNER JOIN `glpi_knowbaseitems`
+                   ON `kb_linked`.`knowbaseitems_id`=`glpi_knowbaseitems`.`id` ";
+
+      if ($item::getType() == KnowbaseItem::getType()) {
+         $id_field = 'knowbaseitems_id';
+         $query .= KnowbaseItem::addvisibilityjoins() . ' WHERE ' .
+            KnowbaseItem::addVisibilityRestrict();
+      } else {
+         $id_field = 'items_id';
+         $restrict = getEntitiesRestrictRequest(
+            '',
+            $item::getTable(),
+            '',
+            '',
+            true
+         );
+         if ($restrict !== '') {
+            $query .= ' INNER JOIN `' . $item->getTable() .
+               '` ON `kb_linked`.`items_id`=`' . $item->getTable() . '`.`id`' .
+               ' WHERE ' . $restrict;
+         }
+      }
+      $query .= " AND `kb_linked`.`$id_field` = '$items_id'";
+
+      if ($sqlfilter) {
+         $query .= " AND ($sqlfilter) ";
+      }
+      $query .= " ORDER BY `itemtype`, `items_id` DESC";
+
+      if ($limit) {
+         $query .= " LIMIT ".intval($start)."," . intval($limit);
+      }
+
```

Comment:
i think, like this pr will target 9.2 (and use current master), maybe we should use the new abstraction provided by remi instead of direct SQL (if possible).